[Kineto] Initialize libkineto profilers during torch init process during pybind set-up

aaronenyeshi opened the pull request:
Summary:
We are planning to lazily initialize CUPTI when profiling is actually performed. Therefore, we need to remove profiler init dependency on CUPTI Callbacks' RESOURCE_CONTEXT_CREATED.

Instead, we can initialize the profilers during torch profiler pybind, ie. THPAutograd_initExtension() and lazily in profilerStep().

Test Plan:
CI and ran internally, see internal diff logs.

Pulled By: aaronenyeshi

albanD reviewed (commented):
Any c++ test we can add to ensure this works smoothly?

aaronenyeshi commented:
> Any c++ test we can add to ensure this works smoothly?

I've added a test in the python test_autograd.py to check that Kineto profilers are initialized after the python bindings are created. Please let me know if it should be moved. Thanks!

davidberard98 commented:
repeating a suggestion I made offline to Aaron - I suggested adding a test like this

https://github.com/pytorch/pytorch/blob/44d0226690b1d3840464fefdb26fb512291b9ecc/test/test_testing.py#L2256-L2258

reasoning - we want to make sure things get initialized immediately; if we add the test as just a normal test, other tests could change the state by causing kineto to get initialized. (Or if we don't need immediate initialization, we probably have something in particular that we expect should trigger initialization, and we would probably want to test that path in isolation)

aaronenyeshi commented:
I've isolated the test, and that actually caught another init bug, it seemed the profiler wasn't registered properly. thanks for the review, this new isolated test is passing:
```
$ python test/test_testing.py -k test_libkineto_profiler_is_initialized
.
----------------------------------------------------------------------
Ran 1 test in 1.906s

OK
```

albanD reviewed (approved):
Thanks!

aaronenyeshi commented:
Fixed the lintrunner issues, and some unrelated ambiguity issues in amdgpu.

aaronenyeshi commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "This break TestCuda::test_lazy_init on ROCm" -c nosignal

huydhn commented:
@pytorchbot revert -m "This break TestCuda::test_lazy_init on ROCm" -c nosignal

aaronenyeshi commented:
@mwootton , @jeffdaily , @malfet - wondering if you have any idea why the library is failing TestCuda::test_lazy_init on ROCm when initializing the libkineto profiler? I am trying to guard the init with `!defined(__HIP__)`, but that doesn't work. Wondering if you have ideas, thank you very much.

jeffdaily commented:
> @mwootton , @jeffdaily , @malfet - wondering if you have any idea why the library is failing TestCuda::test_lazy_init on ROCm when initializing the libkineto profiler? I am trying to guard the init with `!defined(__HIP__)`, but that doesn't work. Wondering if you have ideas, thank you very much.

I was able to reproduce last night and now fresh this morning, same dev host, same container, can't reproduce.  Race condition?

Use the guard `!defined(USE_ROCM)`.  The `__HIP__` macro is defined by the hip-clang compiler, but your source file torch/csrc/autograd/init.cpp will be compiled by the host compiler.  We define USE_ROCM for the entire ROCm build of PyTorch, both host and device sources.

aaronenyeshi committed:
[Kineto] Initialize libkineto profilers during torch init process during pybind set-up (#112623)

Summary:

We are planning to lazily initialize CUPTI when profiling is actually performed. Therefore, we need to remove profiler init dependency on CUPTI Callbacks' RESOURCE_CONTEXT_CREATED.

Instead, we can initialize the profilers during torch profiler pybind, ie. THPAutograd_initExtension() and lazily in profilerStep().

Test Plan:
CI

Ran this locally with dyno gputrace, when turning off init via CUPTI callbacks, I am still able to trigger dyno gputrace combined with D50471345.
  - Verified with p.step()
  - Verified without p.step()
```
$ TEARDOWN_CUPTI=1 buck2 run mode/opt kineto/libkineto/fb/integration_tests:pytorch_resnet_integration_test -c fbcode.platform010_cuda_version="11.4.2"
File changed: fbcode//caffe2/torch/csrc/autograd/init.cpp
File changed: fbcode//caffe2/torch/profiler/__init__.py
Buck UI: https://www.internalfb.com/buck2/5aa4fa84-1fca-4c67-a896-6f2f05a1b8c1
Network: Up: 464KiB  Down: 11MiB  (reSessionID-db8930a4-4747-4f1e-b8bc-43bfe7041502)
Jobs completed: 12. Time elapsed: 1:13.4s.
Cache hits: 0%. Commands: 6 (cached: 0, remote: 0, local: 6)
BUILD SUCCEEDED
INFO:2023-11-01 10:22:14 4156718:4156718 CuptiActivityProfiler.cpp:225] CUDA versions. CUPTI: 14; Runtime: 11040; Driver: 12000
INFO:2023-11-01 10:22:14 4156718:4163224 DynoConfigLoader.cpp:61] Setting communication fabric enabled = 0
INFO:2023-11-01 10:22:14 4156718:4163224 IpcFabricConfigClient.cpp:93] Setting up IPC Fabric at endpoint: dynoconfigclient668f2605-26dd-4bde-9f55-67ef79270ecf status = failed (null)
INFO:2023-11-01 10:22:15 4156718:4163224 DynoConfigLoader.cpp:61] Setting communication fabric enabled = 0
INFO:2023-11-01 10:22:16 Start: ready to do work 4156718
INFO:2023-11-01 10:22:28 step: 0, peak allocated GPU mem: 2.87GB, peak active GPU mem: 2.87GB, peak reserved GPU mem: 2.97GB.
INFO:2023-11-01 10:22:29 step: 20, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:30 step: 40, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:31 step: 60, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:31 step: 80, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:32 step: 100, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:33 step: 120, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:33 step: 140, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:34 4156718:4163224 ConfigLoader.cpp:271] Received config from dyno:

    ACTIVITIES_COMPRESSION_ALGORITHM=GZIP
    REQUEST_TRACE_ID=EB4A32BD79E08846C68A6376D7D652AD
    REQUEST_GROUP_TRACE_ID=EB4A32BD79E08846C68A6376D7D652AD
    ACTIVITIES_DURATION_MSECS=500
    PROFILE_REPORT_INPUT_SHAPES=true
    PROFILE_PROFILE_MEMORY=false
    PROFILE_WITH_STACK=false
    PROFILE_WITH_FLOPS=false
    PROFILE_WITH_MODULES=false
    ACTIVITIES_MANIFOLD_PATH=gpu_traces/tree/traces/dynocli/0/1698859352/127.0.0.1/

INFO:2023-11-01 10:22:34 step: 160, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:35 step: 180, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:35 4156718:4164582 ActivityProfilerController.cpp:142] Received on-demand activity trace request by  profile timestamp = 1698859361379428517
INFO:2023-11-01 10:22:35 4156718:4164582 output_json.cpp:88] Tracing to temporary file /tmp/libkineto_activities_4156718.json
  Log file: /tmp/libkineto_activities_4156718.json
  Trace start time: 2023-11-01 10:22:41  Trace duration: 500ms
  Warmup duration: 5s
  Max GPU buffer size: 128MB
  Enabled activities: cpu_op,user_annotation,gpu_user_annotation,gpu_memcpy,gpu_memset,kernel,external_correlation,cuda_runtime,cuda_driver,cpu_instant_event,python_function,overhead
  Manifold bucket: gpu_traces
  Manifold object: tree/traces/dynocli/0/1698859352/127.0.0.1/libkineto_activities_4156718.json
  Trace compression enabled: 1
  TTL in seconds: 31536000 (365 days)
INFO:2023-11-01 10:22:35 4156718:4164582 CuptiActivityProfiler.cpp:834] Enabling GPU tracing
INFO:2023-11-01 10:22:37 4156718:4164582 CuptiActivityApi.cpp:46]   CUPTI subscriber after reenable:0x7f50ae8f7760
INFO:2023-11-01 10:22:37 4156718:4164582 CuptiActivityProfiler.cpp:873] Tracing starting in 5s
INFO:2023-11-01 10:22:37 4156718:4164582 CuptiActivityProfiler.cpp:878] Tracing will end in 6s
INFO:2023-11-01 10:22:37 step: 200, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:38 step: 220, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:39 step: 240, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:40 step: 260, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:41 step: 280, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
STAGE:2023-11-01 10:22:41 4156718:4164582 CuptiActivityProfiler.cpp:979] Completed Stage: Warm Up
INFO:2023-11-01 10:22:41 4156718:4164582 CuptiActivityProfiler.cpp:988] Tracing started
INFO:2023-11-01 10:22:41 4156718:4164582 CuptiActivityProfiler.cpp:1013] Tracing complete.
INFO:2023-11-01 10:22:41 step: 300, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
STAGE:2023-11-01 10:22:42 4156718:4164582 CuptiActivityProfiler.cpp:1027] Completed Stage: Collection
INFO:2023-11-01 10:22:42 step: 320, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:42 4156718:4164582 CuptiActivityProfiler.cpp:232] Processing 1 CPU buffers
INFO:2023-11-01 10:22:43 step: 340, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:43 4156718:4164582 CuptiActivityProfiler.cpp:262] Processed 1231082 GPU records (52491712 bytes)
INFO:2023-11-01 10:22:43 4156718:4164582 CuptiActivityProfiler.cpp:294] Error info: Out-of-range = 349169, Blocklisted runtime = 577062, Invalid ext correlations = 0, CPU GPU out-of-order = 0, Unexpected CUDA events = 0, CUPTI stopped early? = 0
INFO:2023-11-01 10:22:43 4156718:4164582 CuptiActivityProfiler.cpp:1060] Traces Recorded:
INFO:2023-11-01 10:22:43 4156718:4164582 CuptiActivityProfiler.cpp:1063] PyTorch Profiler: 1 iterations
INFO:2023-11-01 10:22:43 4156718:4164582 output_json.cpp:422] Chrome Trace written to /tmp/libkineto_activities_4156718.json
INFO:2023-11-01 10:22:43 4156718:4164582 output_json.cpp:470] Renamed the trace file to /tmp/libkineto_activities_4156718.json
STAGE:2023-11-01 10:22:43 4156718:4164582 CuptiActivityProfiler.cpp:1050] Completed Stage: Post Processing
INFO:2023-11-01 10:22:43 4156718:4164582 CuptiActivityApi.cpp:391] teardownCupti starting
INFO:2023-11-01 10:22:43 4156718:4165069 CuptiActivityApi.cpp:407]   CUPTI subscriber before finalize:0x7f50ae8f7760
INFO:2023-11-01 10:22:43 4156718:4164224 CuptiCallbackApi.cpp:106]   Calling cuptiFinalize in exit callsite
INFO:2023-11-01 10:22:43 4156718:4165069 CuptiActivityApi.cpp:412] teardownCupti complete
INFO:2023-11-01 10:22:44 step: 360, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:44 step: 380, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:45 step: 400, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:46 step: 420, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:46 step: 440, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:47 step: 460, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:47 4156718:4165068 ManifoldChromeTraceLogger.cpp:137] Uploaded the trace file to Manifold: gpu_traces/tree/traces/dynocli/0/1698859352/127.0.0.1/libkineto_activities_4156718.json.gz
INFO:2023-11-01 10:22:47 4156718:4165068 ManifoldChromeTraceLogger.cpp:139] Check the trace by opening the below URI in your Chrome web browser:
INFO:2023-11-01 10:22:47 4156718:4165068 ManifoldChromeTraceLogger.cpp:141] https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/dynocli/0/1698859352/127.0.0.1/libkineto_activities_4156718.json.gz&bucket=gpu_traces
INFO:2023-11-01 10:22:47 4156718:4165068 ManifoldChromeTraceLogger.cpp:151] Trace upload time: 3997856 us
STAGE:2023-11-01 10:22:47 4156718:4165068 ManifoldChromeTraceLogger.cpp:212] Completed Stage: Manifold Push
INFO:2023-11-01 10:22:48 step: 480, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
INFO:2023-11-01 10:22:48 step: 500, peak allocated GPU mem: 3.10GB, peak active GPU mem: 3.10GB, peak reserved GPU mem: 3.59GB.
...
```

Reviewed By: albanD

Differential Revision: D50894961

Pulled By: aaronenyeshi

aaronenyeshi commented:
@pytorchbot merge

aaronenyeshi commented:
> Use the guard `!defined(USE_ROCM)`. The `__HIP__` macro is defined by the hip-clang compiler, but your source file torch/csrc/autograd/init.cpp will be compiled by the host compiler. We define USE_ROCM for the entire ROCm build of PyTorch, both host and device sources.

@jeffdaily , thank you! That seems to get me past the lazy init issue in ROCM.
